### PR TITLE
Improve stability of focal loss impl

### PR DESCRIPTION
**Pull request by bharrisau**
`torch.exp(neg_logits)` will give a NaN for a large negative value of `neg_logits`. Prevent this by replacing `torch.log1p(torch.exp(neg_logits))` with `torch.logaddexp(neg_logits, 0)`.

`logaddexp(x, y)` is `log(exp(x) + exp(y))`, so `logaddexp(x, 0)` is `log(exp(x) + 1)`. This is equivalent to `log1p(exp(x))` which is just `log(exp(x) + 1)`.

The implementation of `logaddexp` does not return a `NaN` if the operands are very large. It is equivalent to `max(x, y) + log(1 + exp(-1 * abs(x - y)))`. This forces the `exp` to be called with a large negative value instead of a large positive value.

I'm not clear on how this changes the back-propagation behaviour. Also, I think `logaddexp` was only available from pytorch 1.6.0.

### CODE REVIEW

**Comment by bharrisau** [comment]
It seems I've copied the wrong line into the PR. The `0` needs to be a tensor, I just used `neg_logits.new_zeros(1)` instead.

**Comment by rwightman** [comment]
@bharrisau did you run through a full training with reasonable results using this change? I don't want to break PyTorch 1.4 support just yet so will have to figure something out for that.

**Comment by rwightman** [comment]
BTW, wouldn't softplus be appropriate here? torch.nn.functional.softplus ... above a threshold it returns x. I wouldn't have any doubts about gradients because it's used extensively as an activation fn.

I've also been meaning to bring my loss fn up to date with the Tensorflow original. They've refactored it to include label smoothing and it no longer includes the log1p(exp(x) (softplus) term.

**Comment by bharrisau** [comment]
Ok. I was just having issues in LR find. Instead of the loss growing slowly at a high learning rate it would quickly become NaN. After a bit of digging I chased the NaN down to the `exp(large neglogits value)`.

Instead of `softplus(neg_logits)`, this issue suggests `-F.logsigmoid(logits)` (pytorch/pytorch#31856).

The code becoming
```
modulator = torch.exp(gamma * F.logsigmoid(logits) - gamma * targets * logits)
```

```python
targets = torch.tensor([0, 0, 1, 1])
logits = torch.tensor([0.2, 0.8, 0.2, 0.8])
neg_logits = -1.0 * logits
gamma = 2.0

torch.exp(gamma * targets * neg_logits - gamma * torch.log1p(torch.exp(neg_logits)))
> tensor([0.3023, 0.4761, 0.2026, 0.0961])

torch.exp(gamma * F.logsigmoid(logits) - gamma * targets * logits)
> tensor([0.3023, 0.4761, 0.2026, 0.0961])
```

I haven't had a chance to test this `logsigmoid` one out. I used the `logaddexp` and got comparable results. `logsigmoid` is supported on PyTorch 1.4.0. I'll be trying it out later and will probably post a followup with the LR finder curves and training curves of the current method and the `logsigmoid` one.

**Comment by rwightman** [comment]
@bharrisau did you try softplus and run into issues? logsigmoid is implemented by -softplus(-x) in some frameworks, pytorch has a separate impl for each, I'm not sure which one is faster/more stable.

**Comment by bharrisau** [comment]
I just saw that PyTorch has a `threshold` setting for numerical stability
but logsigmoid didn't. Haven't tried with the softplus yet.

**Comment by rwightman** [comment]
@bharrisau I ran some experiments, neither softplus or logsigmoid seemed to be noteably more stable than the current impl. Both were slower. I'm running this on a version of my 'performance' branch with a LOT more changes that haven't been pushed but the loss is similar, just jit scripted.

**Comment by bharrisau** [comment]
No worries. Like I said, the only benefit of the change was when running a
LR finder to pick a good learning rate. I've got a lot less experience than
you on how it relates to the gradients and convergence of the EffDet model,
so you can just close it if it's not going to be an improvement.

**Comment by rwightman** [comment]
@bharrisau I think I'll close this now, it was a wortwhile exploration. Focal loss in general isn't the most stable so the range of working hparams tend to be fairly narrow. I'm currently running training with an updated loss based on the current in automl, it seems to work quite well so I'll likely replace the old one with that when I'm done integrating and testing it with other changes.
